File: backend/app/models/workspace.py
from __future__ import annotations

from datetime import datetime
from uuid import uuid4

from sqlalchemy import Boolean, DateTime, Float, ForeignKey, Integer, String, Text, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column

from app.core.time import utcnow
from app.models.base import Base


class Project(Base):
    __tablename__ = "projects"
    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid4()))
    owner_user_id: Mapped[str] = mapped_column(ForeignKey("users.id"), index=True)
    name: Mapped[str] = mapped_column(String(120))
    description: Mapped[str] = mapped_column(Text, default="")
    github_repo_id: Mapped[int | None] = mapped_column(Integer, nullable=True)
    github_repo_full_name: Mapped[str | None] = mapped_column(String(256), nullable=True)
    github_default_branch: Mapped[str | None] = mapped_column(String(128), nullable=True)
    github_html_url: Mapped[str | None] = mapped_column(String(512), nullable=True)
    security_level: Mapped[str] = mapped_column(String(16), default="standard")
    base_url: Mapped[str | None] = mapped_column(String(512), nullable=True)
    criticality: Mapped[str] = mapped_column(String(16), default="medium")
    scan_options_json: Mapped[str | None] = mapped_column(Text, nullable=True)
    notify_email_default: Mapped[bool] = mapped_column(Boolean, default=True)
    notify_in_app_default: Mapped[bool] = mapped_column(Boolean, default=True)
    auto_scan_on_push: Mapped[bool] = mapped_column(Boolean, default=False)
    auto_scan_branch: Mapped[str | None] = mapped_column(String(128), nullable=True)
    github_webhook_id: Mapped[int | None] = mapped_column(Integer, nullable=True)
    github_webhook_secret: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utcnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utcnow, onupdate=utcnow)


class Scan(Base):
    __tablename__ = "scans"
    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid4()))
    project_id: Mapped[str] = mapped_column(ForeignKey("projects.id"), index=True)
    created_by_user_id: Mapped[str] = mapped_column(ForeignKey("users.id"), index=True)
    target: Mapped[str] = mapped_column(String(512))
    source: Mapped[str] = mapped_column(String(32), default="manual")
    status: Mapped[str] = mapped_column(String(32), default="queued", index=True)
    security_level: Mapped[str] = mapped_column(String(16), default="standard")
    scan_mode: Mapped[str] = mapped_column(String(32), default="rules_only")
    scan_scope: Mapped[str] = mapped_column(String(16), default="full")
    ref: Mapped[str | None] = mapped_column(String(128), nullable=True)
    commit_sha: Mapped[str | None] = mapped_column(String(64), nullable=True)
    commit_short: Mapped[str | None] = mapped_column(String(16), nullable=True)
    commit_message: Mapped[str | None] = mapped_column(String(512), nullable=True)
    commit_author: Mapped[str | None] = mapped_column(String(256), nullable=True)
    git_history_json: Mapped[str | None] = mapped_column(Text, nullable=True)
    share_token: Mapped[str | None] = mapped_column(String(64), unique=True, nullable=True, index=True)
    progress_json: Mapped[str | None] = mapped_column(Text, nullable=True)
    eta_seconds: Mapped[int | None] = mapped_column(Integer, nullable=True)
    error_message: Mapped[str | None] = mapped_column(Text, nullable=True)
    summary_json: Mapped[str | None] = mapped_column(Text, nullable=True)
    risk_summary_json: Mapped[str | None] = mapped_column(Text, nullable=True)
    notify_email: Mapped[bool] = mapped_column(Boolean, default=True)
    notify_in_app: Mapped[bool] = mapped_column(Boolean, default=True)
    cancel_requested: Mapped[bool] = mapped_column(Boolean, default=False)
    started_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    finished_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utcnow, index=True)


class Finding(Base):
    __tablename__ = "findings"
    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid4()))
    scan_id: Mapped[str] = mapped_column(ForeignKey("scans.id"), index=True)
    engine: Mapped[str] = mapped_column(String(32), index=True)
    rule_id: Mapped[str | None] = mapped_column(String(256), nullable=True)
    vuln_family: Mapped[str] = mapped_column(String(32), default="other", index=True)
    cwe: Mapped[str | None] = mapped_column(String(32), nullable=True)
    owasp_category: Mapped[str | None] = mapped_column(String(64), nullable=True)
    severity: Mapped[str] = mapped_column(String(16), default="info", index=True)
    title: Mapped[str] = mapped_column(String(512))
    message: Mapped[str | None] = mapped_column(Text, nullable=True)
    file_path: Mapped[str | None] = mapped_column(String(1024), nullable=True)
    line_start: Mapped[int | None] = mapped_column(Integer, nullable=True)
    line_end: Mapped[int | None] = mapped_column(Integer, nullable=True)
    snippet: Mapped[str | None] = mapped_column(Text, nullable=True)
    risk_score: Mapped[float] = mapped_column(Float, default=0.0)
    status: Mapped[str] = mapped_column(String(16), default="open", index=True)
    fingerprint: Mapped[str | None] = mapped_column(String(64), nullable=True, index=True)
    ai_verdict: Mapped[str | None] = mapped_column(String(32), nullable=True)
    ai_rationale: Mapped[str | None] = mapped_column(Text, nullable=True)
    countermeasures_json: Mapped[str | None] = mapped_column(Text, nullable=True)
    raw_json: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utcnow)


class FindingSuppression(Base):
    __tablename__ = "finding_suppressions"
    __table_args__ = (UniqueConstraint("project_id", "fingerprint", name="uq_project_fingerprint"),)
    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid4()))
    project_id: Mapped[str] = mapped_column(ForeignKey("projects.id"), index=True)
    fingerprint: Mapped[str] = mapped_column(String(64), index=True)
    reason: Mapped[str | None] = mapped_column(String(512), nullable=True)
    created_by_user_id: Mapped[str] = mapped_column(ForeignKey("users.id"), index=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utcnow)


class AppNotification(Base):
    __tablename__ = "app_notifications"
    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid4()))
    user_id: Mapped[str] = mapped_column(ForeignKey("users.id"), index=True)
    scan_id: Mapped[str | None] = mapped_column(ForeignKey("scans.id"), nullable=True)
    title: Mapped[str] = mapped_column(String(256))
    body: Mapped[str] = mapped_column(Text, default="")
    read_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utcnow, index=True)


class GitHubConnection(Base):
    __tablename__ = "github_connections"
    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid4()))
    user_id: Mapped[str] = mapped_column(ForeignKey("users.id"), unique=True, index=True)
    github_user_id: Mapped[int] = mapped_column(Integer, index=True)
    github_login: Mapped[str] = mapped_column(String(128))
    avatar_url: Mapped[str | None] = mapped_column(String(512), nullable=True)
    token_encrypted: Mapped[str] = mapped_column(Text)
    scopes: Mapped[str] = mapped_column(String(256), default="")
    needs_reauth: Mapped[bool] = mapped_column(Boolean, default=False)
    connected_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utcnow)
    revoked_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

File: frontend/src/lib/workspace.ts
export type Project = {
  id: string;
  name: string;
  description: string;
  github_repo_id: number | null;
  github_repo_full_name: string | null;
  github_default_branch: string | null;
  github_html_url: string | null;
  security_level?: string;
  base_url?: string | null;
  criticality?: string;
  notify_email_default?: boolean;
  notify_in_app_default?: boolean;
  auto_scan_on_push?: boolean;
  auto_scan_branch?: string | null;
  created_at: string;
  updated_at: string;
};

export type ScanProgressLog = {
  t?: string;
  msg: string;
};

export type ScanProgress = {
  phase?: string;
  label?: string;
  percent?: number;
  eta_remaining_seconds?: number | null;
  logs?: ScanProgressLog[];
  findings_so_far?: number;
};

export type GitCommit = {
  sha: string;
  short: string;
  message: string;
  author: string;
  date: string;
};

export type Scan = {
  id: string;
  project_id: string;
  project_name?: string | null;
  target: string;
  source: string;
  status: string;
  security_level?: string;
  scan_mode?: string;
  scan_scope?: string;
  ref?: string | null;
  commit_sha?: string | null;
  commit_short?: string | null;
  commit_message?: string | null;
  commit_author?: string | null;
  git_history?: GitCommit[];
  share_token?: string | null;
  shared?: boolean;
  progress?: ScanProgress | null;
  eta_seconds?: number | null;
  error_message?: string | null;
  cancel_requested?: boolean;
  summary?: {
    findings_count?: number;
    engine?: string;
    note?: string;
    by_severity?: Record<string, number>;
    by_family?: Record<string, number>;
    ai_status?: string;
    ai_report?: string | null;
    max_risk?: number;
    commit_short?: string;
    commit_sha?: string;
    engines?: Array<Record<string, unknown>>;
    options?: Record<string, unknown>;
    policy_failed?: boolean;
    fail_severity?: string;
  } | null;
  risk_summary?: { max_risk?: number; open_findings?: number; by_severity?: Record<string, number> } | null;
  notify_email?: boolean;
  notify_in_app?: boolean;
  options?: {
    engines?: string[];
    path_excludes?: string[];
    fail_severity?: string;
    code_review?: boolean;
  } | null;
  started_at?: string | null;
  finished_at?: string | null;
  created_at: string;
};

export type FindingStatus = "open" | "triage" | "fixed" | "false_positive";

export type Finding = {
  id: string;
  scan_id: string;
  engine: string;
  rule_id?: string | null;
  vuln_family: string;
  cwe?: string | null;
  owasp_category?: string | null;
  severity: string;
  title: string;
  message?: string | null;
  file_path?: string | null;
  line_start?: number | null;
  line_end?: number | null;
  snippet?: string | null;
  risk_score: number;
  status: string;
  fingerprint?: string | null;
  ai_verdict?: string | null;
  ai_rationale?: string | null;
  countermeasures?: Array<{ title?: string; steps?: string[] }> | null;
};

export type GitHubStatus = {
  configured: boolean;
  missing?: string[];
  env_file_found?: boolean;
  connected: boolean;
  needs_reauth?: boolean;
  github_login: string | null;
  avatar_url: string | null;
  scopes?: string | null;
  connected_at: string | null;
};

export type GitHubRepo = {
  id: number;
  full_name: string;
  name: string;
  private: boolean;
  default_branch: string | null;
  html_url: string | null;
  description: string;
  updated_at?: string | null;
};

export type GitRef = { name: string; type: "branch" | "tag" | string };

export type WorkspaceDashboard = {
  totals: { projects: number; scans_this_week: number; open_findings?: number };
  github: GitHubStatus;
  recent_scans: Scan[];
  projects: Project[];
  charts?: {
    runs: Array<{
      id: string;
      project_id: string;
      project_name?: string | null;
      label: string;
      status: string;
      created_at: string;
      findings_count: number;
      by_severity: Record<string, number>;
      by_family: Record<string, number>;
    }>;
    open_by_severity: Record<string, number>;
    open_by_severity_by_project?: Record<string, Record<string, number>>;
    open_by_family: Record<string, number>;
    open_by_engine: Record<string, number>;
    scan_outcomes: Record<string, number>;
  };
};
